dag_in_context/src/pretty_print.rs

```diff
@@ -0,0 +1,831 @@
+    }
+
+    pub fn to_ast(&self) -> String {
+        match self {
+            BaseType::IntT => "intt()".into(),
+            BaseType::BoolT => "boolt()".into(),
+            BaseType::StateT => "statet()".into(),
+            BaseType::PointerT(ptr) => format!("pointert({})", BaseType::to_ast(ptr)),
+            BaseType::FloatT => "floatt()".into(),
+        }
+    }
+
+    pub fn abbrev(&self) -> String {
+        match self {
+            BaseType::IntT => "i".into(),
+            BaseType::BoolT => "b".into(),
+            BaseType::StateT => "s".into(),
+            BaseType::PointerT(ptr) => format!("ptr{}", &ptr.abbrev()),
+            BaseType::FloatT => "f".into(),
+        }
+    }
+}
+
+impl Type {
+    pub fn pretty(&self) -> String {
+        let (term, termdag) = self.to_egglog();
+        let expr = termdag.term_to_expr(&term);
+        expr.to_sexp().pretty()
+    }
+
+    pub fn to_ast(&self) -> String {
+        match self {
+            Type::Base(t) => format!("base({})", BaseType::to_ast(t)),
+            Type::TupleT(vec_ty) => {
+                if vec_ty.is_empty() {
+                    return "emptyt()".into();
+                }
+                let vec_ty_str = vec_ty
+                    .iter()
+                    .map(BaseType::to_ast)
+                    .collect::<Vec<_>>()
+                    .join(", ");
+                format!("tuplet!({vec_ty_str})")
+            }
+            Type::Unknown => panic!("found unknown in to_ast"),
+            Type::Symbolic(str) => format!("{}.clone()", str),
+        }
+    }
+
+    // give a abbreviated name of type, ex: a tuple of Int Int State become tpl_iis
+    pub fn abbrev(&self) -> String {
+        match self {
+            Type::Base(base) => format!("base_{}", base.abbrev()),
+            Type::TupleT(vec) => {
+                let vec_ty_str = vec
+                    .iter()
+                    .map(|bt| bt.abbrev())
+                    .collect::<Vec<_>>()
+                    .join("_");
+                format!("tpl_{}", vec_ty_str)
+            }
+            Type::Unknown => "unknown".into(),
+            Type::Symbolic(str) => str.into(),
+        }
+    }
+}
+
+impl BinaryOp {
+    pub fn to_ast(&self) -> String {
+        use schema::BinaryOp::*;
+        // the same as schema_helper's
+        match self {
+            Add => "add",
+            Sub => "sub",
+            Mul => "mul",
+            Div => "div",
+            Eq => "eq",
+            LessThan => "less_than",
+            GreaterThan => "greater_than",
+            LessEq => "less_eq",
+            GreaterEq => "greater_eq",
+            And => "and",
+            Or => "or",
+            PtrAdd => "ptradd",
+            Load => "load",
+            Print => "tprint",
+            Free => "free",
+            FGreaterEq => "f_greater_eq",
+            FGreaterThan => "f_greater_than",
+            FLessEq => "f_less_eq",
+            FLessThan => "f_less_than",
```

Comment:
are these in ast.rs? Do you need to add them?

Reply:
yes I thought it was added